### PR TITLE
Make line/continuum fitting work in the mask case (and other mask-related tests)

### PR DESCRIPTION
**Pull request by eteq**
This is the PR I promised in #516. It fixes the line-fitting machinery to ignore any masked parts of the spectrum when doing a fit.

Note this also includes a couple of tests I added in process of the above to ensure the masks behave correctly for arithmetic and spectral regions.

cc @brechmos-stsci @camipacifici @nmearl 

(This may particularly benefit from a review by @brechmos-stsci who did a fair amount of the fitting stuff)

### CODE REVIEW

File: specutils/fitting/fitmodels.py [hunk 1]
```diff
@@ -284,10 +284,12 @@ def fit_lines(spectrum, model, fitter=fitting.LevMarLSQFitter(),
         Fitter instance to be used when fitting model to spectrum.
     exclude_regions : list of `~specutils.SpectralRegion`
         List of regions to exclude in the fitting.
-    weights : list or 'unc', optional
+    weights : array-like or 'unc', optional
         If 'unc', the unceratinties from the spectrum object are used to
-        to calculate the weights. If list/ndarray, represents the weights to
-        use in the fitting.
+        to calculate the weights. If array-like, represents the weights to
+        use in the fitting.  Not that if a mask is present on the spectrum, it
```
Comment (keflavich) [inline specutils/fitting/fitmodels.py]:
```suggestion
        use in the fitting.  Note that if a mask is present on the spectrum, it
```

File: specutils/fitting/fitmodels.py [hunk 2]
```diff
@@ -418,40 +423,49 @@ def _fit_lines(spectrum, model, fitter=fitting.LevMarLSQFitter(),
     #
 
     # In this case the window defines the area around the center of each model
+    window_indices = None
     if window is not None and isinstance(window, (float, int)):
         center = model.mean
-        indices = np.nonzero((spectrum.spectral_axis >= center-window) &
+        window_indices = np.nonzero((spectrum.spectral_axis >= center-window) &
```
Comment (keflavich) [inline specutils/fitting/fitmodels.py]: probably should change indentation of subsequent line?

File: specutils/fitting/fitmodels.py [hunk 3]
```diff
@@ -418,40 +423,49 @@ def _fit_lines(spectrum, model, fitter=fitting.LevMarLSQFitter(),
...
-                raise Exception("Bad selected region.")
-            extracted_regions = extract_region(spectrum, window)
-            dispersion, flux = _combined_region_data(extracted_regions)
-            dispersion = dispersion * dispersion_unit
-            flux = flux * flux_unit
-        except ValueError as e:
-            return
+        idx1, idx2 = window.bounds
+        if idx1 == idx2:
+            raise Exception("Bad selected region.")
```
Comment (keflavich) [inline specutils/fitting/fitmodels.py]: don't raise a base exception - this could be an `IndexError`?

File: specutils/fitting/fitmodels.py [hunk 4]
```diff
@@ -418,40 +423,49 @@ def _fit_lines(spectrum, model, fitter=fitting.LevMarLSQFitter(),
...
+
+        # HACK WARNING! This uses the extract machinery to create a set of
+        # indices by making an "index spectrum"
+        # TODO: really the spectral region machinery should have the power
+        # to create a mask, and we'd just use that...
+        index_spectrum = Spectrum1D(spectral_axis=spectrum.spectral_axis,
+                                    flux=np.arange(spectrum.flux.size).reshape(spectrum.flux.shape)*u.Jy)
```
Comment (keflavich) [inline specutils/fitting/fitmodels.py]: is the `Jy` unit addition needed?  That adds a small performance hit and is weird for what should be an integer array to begin with (on which note, should this be `arange(..., dtype='int')`?)

File: specutils/tests/spectral_examples.py [hunk 5]
```diff
@@ -71,22 +76,29 @@ def __init__(self):
...
 
         self._flux_e4 = (self.base_flux + 400 * np.random.random(self.base_flux.shape)) * 1000000
         self._s1_AA_nJy_e4 = Spectrum1D(spectral_axis=self.wavelengths_AA * u.AA,
                                         flux=self._flux_e4 * u.nJy)
 
 
+        #
+        # Create one spectrum like 1 but with a mask
+        #
+        self._s1_um_mJy_e1_masked = copy(self._s1_um_mJy_e1)  # SHALLOW copy
```
Comment (keflavich) [inline specutils/tests/spectral_examples.py]: maybe comment why shallow is emphasized here?  It's true, but does it matter in context?

File: specutils/tests/test_fitting.py [hunk 6]
```diff
@@ -595,3 +595,55 @@ def test_spectrum_from_model():
                               1.5319218, 1.06886794, 0.71101768, 0.45092638, 0.27264641])
 
     assert np.allclose(spectrum_gaussian.flux.value[::10], flux_expected, atol=1e-5)
+
+
+def test_masking():
+    """
+    Test fitting spectra with masks
+    """
+    x, y = double_peak()
```
Comment (keflavich) [inline specutils/tests/test_fitting.py]: avoid single-character variable names.... (my preference, not an official style recommendation, afaik)

File: specutils/fitting/fitmodels.py (hunk 3 above)
Comment (eteq) [reply specutils/fitting/fitmodels.py]: pre-existing before this PR, but I agree!

File: specutils/fitting/fitmodels.py (hunk 4 above)
Comment (eteq) [reply specutils/fitting/fitmodels.py]: The reasoning for this is that `Spectrum1D` expects flux inputs to be quantities - i.e., you can't give something that's unitless.  The choice of Jy is arbitrary, as really any quantity will do, but it seemed safest to use an actual flux. But you're right it makes sense to keep it as an `int` so I can switch to the explicit `u.Quantity(arr, u.Jy, dtype=int)`

As the "HACK WARNING" indicates this is defintely a hack, though I think it's probably better just to try to do the "TODO" ASAP instead beyond optimizing this any more.

File: specutils/tests/test_fitting.py (hunk 6 above)
Comment (eteq) [reply specutils/tests/test_fitting.py]: I think it's also in several of the default linters, so it's quasi-official... regardless, good point will fix

**Comment by eteq** [comment]
I think I've addressed everything, so this should be good to go once the tests pass if @keflavich or @nmearl are willing to approve/merge it.

**Comment by nmearl** [comment]
Looks like there's a hold over from the recent `MexicanHat` model name switch.

**Comment by eteq** [comment]
@nmearl  - Re the mexicanhat failure:   I don't think it's related to this PR. I was about to added it but I realized we should treat it separately because I think we may want to leave the "old" name in for the Astropy 3.x compatible version.  So I made a separate PR #564 to address that and propose this be merged separately.

**Comment by nmearl** [comment]
Just to clarify: this PR would be in for the astropy 3.x release, but the PR #564 would be included in the astropy 4.x release?

**Comment by eteq** [comment]
@nmearl - yep, exactly!

**Comment by eteq** [comment]
(since the test failure is unrelated I think this is good to go, and a release blocker)

**Comment by pllim** [comment]
This PR actually introduced a bug that is being fixed at https://github.com/astropy/specutils/pull/1164 . FYI

**Pull request merged**